### ISSUE TITLE
make_glm_oracle.py on transformers 5.5.0 produces an oracle the engine scores 25/32 against (expected 32/32)

### ISSUE THREAD

**Issue by KingIcyCreamProjects** [JustVugg/colibri#281]
Regenerating the tiny oracle from scratch on a current stack — `tools/make_glm_oracle.py` with transformers **5.5.0** / torch 2.11.0 (Windows 11, Python 3.13) — and then running the documented self-test:

```
SNAP=./glm_tiny TF=1 ./glm 64 16 16
PREFILL (teacher-forcing) C vs oracle: 25/32 positions | 2505.9 pos/s
```

Expected 32/32 per setup.sh. The engine binary is bit-identical before/after on my side and scores as documented against the *shipped* `ref_glm.json`, so this looks like implementation drift in transformers' `GlmMoeDsa*` between the version the oracle was originally generated with and 5.5.0 (router/indexer details or numerics), not an engine regression. Worth pinning a known-good transformers version in the oracle script's docstring (or regenerating the shipped oracle and noting the version) so `setup.sh`'s 32/32 stays reproducible for new contributors.

**Comment by JustVugg** [comment]
Thanks — this looks like a transformers 5.5.0 behavior change in the reference path, not an engine regression (the engine is token-exact against oracles built on the pinned versions). `make_glm_oracle.py` builds the reference greedy trace via `GlmMoeDsaForCausalLM`; if 5.5.0 changed a default (RoPE, attention scaling, or the MoE routing tie-break), the oracle drifts and the 32/32 no longer holds even though the engine is unchanged. Two asks: (1) which transformers version gave you 32/32 before? (2) can you diff the first divergent token's logits? I'll pin the known-good transformers range in the tool's header and add a version check that warns on mismatch. Keeping open.

**Comment by KingIcyCreamProjects** [comment]
Bisected it the direction is inverted: **no older version gives 32/32. Current transformers matches your shipped oracle; 5.5.0 is the drifted one.**

Regenerated the oracle under 14 transformers versions (same seed, torch 2.11.0, CPU):

| transformers | result |
|---|---|
| 5.2.0 – 5.10.4 (incl. my 5.5.0) | drifted → engine 25/32 |
| 5.11.0 – 5.14.0 | identical to shipped `ref_glm.json` → engine 32/32 |

Weights come out hash-identical on every version, so it's purely a forward-pass change.

**(2) Logits diff** (same weights, same input, 5.14.0 vs 5.5.0): first divergence at position 5, and it's a near-tie — 5.14.0 puts token 34 ahead of 197 by 0.0025 logits, 5.5.0 puts 197 ahead by 0.022. Hidden-state deltas start tiny at layer 1 (~3e-2) and blow up at the MoE layers (0.4 → 1.5): a small attention-level difference, amplified by router flips.

**Root cause:** in ≤5.10.4 the main MLA attention applied split-half (Llama-style) RoPE; 5.11.0 switched it to DeepSeek-style interleaved (`apply_rotary_pos_emb_interleave`) — which is what the engine implements. Verified causally: 5.10.4 with just that one function patched to the interleaved math regenerates an oracle byte-identical to the shipped one.

So the pin is a floor, not a range: `transformers>=5.11`. Happy to PR the version check if useful.

> Thanks — this looks like a transformers 5.5.0 behavior change in the reference path, not an engine regression (the engine is token-exact against oracles built on the pinned versions). `make_glm_oracle.py` builds the reference greedy trace via `GlmMoeDsaForCausalLM`; if 5.5.0 changed a default (RoPE, attention scaling, or the MoE routing tie-break), the oracle drifts and the 32/32 no longer holds even though the engine is unchanged. Two asks: (1) which transformers version gave you 32/32 before? (2) can you diff the first divergent token's logits? I'll pin the known-good transformers range in the tool's header and add a version check that warns on mismatch. Keeping open.

**Comment by lornecodes** [comment]
Two datapoints for this from a fresh Linux setup today (referred here from #108):

**1. A version datapoint that may bisect this:** oracle regeneration on **transformers 5.14.1** (torch 2.13.0+cpu, gcc 14.2, Debian, python 3.13) produces a clean **TF 32/32** against the freshly written `ref_glm.json` — first try, greedy path untested but teacher-forcing exact. If the 25/32 here was on transformers 5.5.0, the mismatch may be version-specific (GlmMoeDsa numerics or a fixed bug between 5.5.0 → 5.14.x) rather than a converter/engine problem.

**2. A stale-artifact failure mode worth ruling out:** `make_glm_oracle.py` crashes if `c/glm_tiny/` doesn't exist (safetensors `save_file` → "No such file or directory" on its tempfile; needs a `mkdir -p` before save). The flip side of that bug: the script writes `glm_tiny/` weights and `ref_glm.json` as **separate steps**, so any crash or interrupt between them leaves a *mismatched pair* — old weights with a new reference (or vice versa) — which would produce exactly a stable, reproducible partial match like 25/32 with no other symptom. Worth checking whether the 25/32 runs were against a `glm_tiny/` regenerated into a directory that already held earlier shards, and whether deleting both and regenerating atomically changes the score.

Suggested hardening (can PR): `out.mkdir(parents=True, exist_ok=True)` + write `ref_glm.json` first to a temp name and rename only after the weights save succeeds.

— Peter Groom, Dawn Field Institute

**Comment by JustVugg** [comment]
Fixed on `dev` — **#322** (@woolcoxm) is merged.

`make_glm_oracle.py` now hard-fails on `transformers < 5.11.0` instead of quietly building a bad oracle. Your diagnosis was the whole fix: those versions apply split-half (Llama-style) RoPE in `GlmMoeDsa*`, while GLM-5.2's MLA uses interleaved (DeepSeek-style) RoPE — which is what the C engine implements. The weights come out identical; only the forward pass differs. So there was no safe "partial" run: an old transformers silently produced an invalid `ref_glm.json` and the engine scored 25/32 against a reference that was itself wrong.

The gate runs **before** the heavy `from transformers import ...` (which can reset the in-memory `__version__`) and reads the installed distribution metadata via `importlib.metadata`, so it can't be fooled by a lazy-loading reset.

Hard-fail rather than warn was the right call, and it's the same principle we've been applying all week: a check that lets a broken run continue is worse than no check — it just moves the confusion downstream. Thank you for tracking this one down; a wrong oracle is the most expensive kind of bug because everything measured against it is also wrong.

**Comment by woolcoxm** [comment]
there was something firing off after the version has been nullified. this caused the script to fire off no matter what, added a hard stop to make sure the proper package is installed before the program will run, the version was literally being stripped the line before the version check.
